cpu/sam3x8e/periph/pwm.c

```diff
@@ -0,0 +1,258 @@
+ * @brief       CPU specific low-level PWM driver implementation for the SAM3X8E
+ *
+ * @author      Andreas "Paul" Pauli" <andreas.pauli@haw-hamburg.de>
+ *
+ * @}
+ */
+#include <stdint.h>
+#include "board.h"          /* -> cpu.h ??? */
+#include "periph/pwm.h"     /* "driver", interface of pwm.c ONLY */
+
+/*
+ * guard file in case no PWM device is defined, all #define'd in "periph_conf.h"
+ */
+#if PWM_NUMOF && (PWM_0_EN || PWM_1_EN || PWM_2_EN || PWM_3_EN)
+
+#define ERR_INIT_MODE           (-1)
+#define ERR_INIT_BWTH           (-2)
+#define ERR_SET_CHAN            (-1)
+
+// Disable IO on PC2 and set Peripheral device B (PWM0L)
+#define PWM_0_PIN_SET                           \
+    PIOC->PIO_PDR |= PIO_PDR_P2;                \
+    PIOC->PIO_ABSR |= PWM_0_PIN;
+
+/*!
+ * Disable GPIO and set peripheral A/B ("0/1") for pwm channel pins.
+ *
+ * \todo { Find out why simple assignment '=' isn't working although
+ * registers are Read only}
+ * \todo { Extend macro use }
+ */
+#define PWM_0_CHAN_PIN_SET() {                          \
+        PIOC->PIO_PDR  |= PIO_PDR_P21;                  \
+        PWM_0_PORT_CH0->PIO_ABSR |= PWM_0_PIN_CH0;      \
+        PIOC->PIO_PDR  |= PIO_PDR_P22;                  \
+        PWM_0_PORT_CH1->PIO_ABSR |= PWM_0_PIN_CH1;      \
+        PIOC->PIO_PDR  |= PIO_PDR_P23;                  \
+        PWM_0_PORT_CH2->PIO_ABSR |= PWM_0_PIN_CH2;      \
+        PIOC->PIO_PDR  |= PIO_PDR_P24;                  \
+        PWM_0_PORT_CH3->PIO_ABSR |= PWM_0_PIN_CH3;      \
+    }
+
+/*!
+ * Only handles one device with max. four channels.
+ *
+ * To ease initialisation the four channels (4-7) are controlled by a sync
+ * group. CH0 is master of sync group and inherently selected.
+ *
+ * - Selecting clock and mode.
+ * - Set period and other comparator paramters.
+ * - En/Disable.
+ *
+ * Duty-cycle update is selected individualy.
+ *
+ * \warning {Although channel 0 isn't routed to a pin, controlling th sync
+ * group means it MUST NOT used for other purposes, Channels 1-3 are still
+ * free usable}
+ * \todo { Use PWM_0_CHANNELS for selective initialisation }
+ */
+int pwm_init (pwm_t dev, pwm_mode_t mode,
+              unsigned int frequency,
+              unsigned int resolution)
+{
+
+    int32_t retval = ERR_INIT_MODE;      /* Worst/First case */
+    uint32_t pwm_clk = 0;                //!< Desired/real pwm_clock //
+    uint32_t diva = 1;                   //!< Candidate for 8bit prescaler //
+    uint32_t prea = 0;                   //!< Candidate for clock select //
+
+    Pwm* pwm;
+    PwmCh_num* chan;
+
+    /*
+     * Mode check.
+     * Involves things that shold be in "periph_conf.h".
+     * !Junk!Just wasting memory an a few cpu cycles for now. ;/
+     */
+    switch ( mode) {
+    case PWM_LEFT:
+        //!\todo
+        break;
+    case PWM_CENTER:
+        //!\todo
+        break;
+    default:
+        //retval = ERR_INIT_BWTH;
+        // Don't like :/
+        return ERR_INIT_BWTH;
+    }
```

Comment:
I tend to have a "completely useless" switch-block to (i) be prepared for further implementations and (ii) highlight the lack so one immediately sees what's missing.